repl: Allow fn parenthesis across multiple lines

tomgco opened the pull request:
This fixes #4060, which breaks due to the error: `missing ) after argument list` and is not treated as recoverable.

cjihrig commented on test/parallel/test-repl-missing-parenthesis-recoverable.js:
```diff
@@ -0,0 +1,33 @@
+ * This is a regression test for https://github.com/nodejs/node/issues/4060
+ */
+require('../common');
+var assert = require('assert');
```
Can you replace the `var`s with `const` and `let` as appropriate.

cjihrig commented on test/parallel/test-repl-missing-parenthesis-recoverable.js:
```diff
@@ -0,0 +1,33 @@
+ */
+require('../common');
+var assert = require('assert');
+
```
No need for the blank line.

cjihrig commented on test/parallel/test-repl-missing-parenthesis-recoverable.js:
```diff
@@ -0,0 +1,33 @@
+
+var spawn = require('child_process').spawn;
+// use -i to force node into interactive mode, despite stdout not being a TTY
+var args = [ '-i' ];
```
No need for the whitespace inside the brackets.

Fishrock123 replied:
I *think* `make jslint` should catch this.

cjihrig commented on test/parallel/test-repl-missing-parenthesis-recoverable.js:
```diff
@@ -0,0 +1,33 @@
+  out += c;
+});
+
+child.stdout.on('end', function() {
```
Can you wrap the callback in `common.mustCall()`.

cjihrig commented on test/parallel/test-repl-missing-parenthesis-recoverable.js:
```diff
@@ -0,0 +1,33 @@
+
+child.stdout.on('end', function() {
+  assert(expectOut.test(out));
+  console.log('ok');
```
Can you remove the `console.log()`.

cjihrig commented:
LGTM with comments and pending CI

tomgco commented:
> LGTM with comments and pending CI

Thanks, will update - I copied most of the boilerplate from another test, shall I go through (in another commit) and retrofit those tests to match this format?

Fishrock123 commented:
I think this is a feature addition, Tagging as such.

cjihrig commented:
> shall I go through (in another commit) and retrofit those tests to match this format?

Thanks, but that's OK. We generally apply these types of changes to old code when it needs to be modified for some other reason.

tomgco commented:
> Thanks, but that's OK. We generally apply these types of changes to old code when it needs to be modified for some other reason.

Ok, @cjihrig @Fishrock123 I have updated the test case addressing your issues. `make jslint` now passes locally for me.

cjihrig commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/1169/

jasnell commented:
CI is green. LGTM

thefourtheye commented on test/parallel/test-repl-missing-parenthesis-recoverable.js:
```diff
@@ -0,0 +1,31 @@
+// Match '...' as well since it marks a multi-line statement
+const expectOut = /^> ... undefined\n/;
+
+child.stderr.setEncoding('utf8');
```
Is this necessary? I thought the default encoding itself is utf8 only.

thefourtheye commented on test/parallel/test-repl-missing-parenthesis-recoverable.js:
```diff
@@ -0,0 +1,31 @@
+const expectOut = /^> ... undefined\n/;
+
+child.stderr.setEncoding('utf8');
+child.stderr.on('data', (c) => {
```
Unused parameter `c`.

cjihrig replied:
You can change this to just `child.stderr.on('data', common.fail);`

princejwesley commented on lib/repl.js:
```diff
@@ -1140,7 +1140,7 @@ function isRecoverableError(e, self) {
       return true;
     }
-    return /^(Unexpected end of input|Unexpected token)/.test(message);
+    return /^(Unexpected (end of input|token)|missing \) after)/.test(message);
```
`/^(?:Unexpected (?:end of input|token)|missing \) after)/` (No need to capture groups)

tomgco commented:
@cjihrig @jasnell I have updated the commit with your suggested changes! Shall we rerun the CI as we are no longer using capture group in the regex?

Also, due to the size of the Regex I had to break the test onto another line (https://github.com/nodejs/node/pull/4585/files#diff-b13d72249263845d8e8341db0426f9d3R1144) Is that OK? What should we do in these situations when the Regex's become too large? Use `RegExp`?

cjihrig commented:
@tomgco It's OK to split across lines. However, in this case you could pull the regex out of the function into a separate variable. I think you can gain some performance that way. See https://github.com/tomgco/node/blob/repl-fix/lib/repl.js#L605-L607

tomgco commented:
> @tomgco It's OK to split across lines. However, in this case you could pull the regex out of the function into a separate variable. I think you can gain some performance that way. See https://github.com/tomgco/node/blob/repl-fix/lib/repl.js#L605-L607

Excellent, Shall we leave that as another issue and tag it with good first contribution?

cjihrig commented:
I'd just take care of it here. One less open issue. One less PR. Less code churn.

tomgco commented:
> I'd just take care of it here. One less open issue. One less PR. Less code churn.

I agree it wouldn't be difficult to do here, I guess that this could be more of a discussion around making it easier to onboard new developers to help out. As a lot of the current "Good first contribution" labels are doc changes or are quite complicated, considerable chunks of work.

I will fix this up here and find a suitable place to open the discussion.

tomgco commented:
@cjihrig Fixed up, hopefully all good now.

cjihrig commented on lib/repl.js:
```diff
@@ -605,6 +605,8 @@ ArrayStream.prototype.write = function() {};
 const requireRE = /\brequire\s*\(['"](([\w\.\/-]+\/)?([\w\.\/-]*))/;
 const simpleExpressionRE =
     /(([a-zA-Z_$](?:\w|\$)*)\.)*([a-zA-Z_$](?:\w|\$)*)\.?$/;
+const recoverableErrosRE =
```
Is `Erros` a typo?

cjihrig replied:
@tomgco ping on this comment.

princejwesley commented:
@kunalspathak @Fishrock123 @cjihrig

REPL is tightly coupled with v8 syntax error messages. Node REPL with chakra engine may cause issues.

kunalspathak commented:
That's right. Syntax error messages differ in v8 and chakra. See https://github.com/Microsoft/node/commit/0faa5df2b9ac70157208efa90cbcde2526221b37 for examples.

thefourtheye commented:
https://github.com/nodejs/node/pull/4765#issuecomment-174093266

> In some day we will write codes like that...

>     if (engine === 'v8') {
        do_stuff_for_v8();
    } else if (engine === 'chakra') {
        do_stuff_for_chakra();
    }
and we all gonna hate about it.

Sigh!

tomgco commented:
> Sigh!

Lets not stop this from landing :D But I guess it is another subsystem that would need to be addressed for VM neutrality https://github.com/nodejs/roadmap/issues/54

thefourtheye commented on test/parallel/test-repl-missing-parenthesis-recoverable.js:
```diff
@@ -0,0 +1,27 @@
+'use strict';
+/*
```
JavaScript the Good Parts says

> The `/* */` form of block comments came from a language called PL/I. PL/I chose
those strange pairs as the symbols for comments because they were unlikely to occur
in that language’s programs, except perhaps in string literals. In JavaScript, those
pairs can also occur in regular expression literals, so block comments are not safe for
commenting out blocks of code

I am neutral about this though.

cjihrig replied:
We've got block comments all through the codebase. However, for a single line comment, it seems unnecessary. Either way, not a deal breaker.

jasnell commented:
LGTM

cjihrig commented:
@jasnell there is one thing I'm waiting to hear back from @tomgco on. Other than that, LGTM

thefourtheye commented:
LGTMT, with the typo fix suggested by @cjihrig

tomgco commented:
> @jasnell there is one thing I'm waiting to hear back from @tomgco on. Other than that, LGTM

Sorry, totally missed that comment. I will update the typo and the `/* */` comment when I get home, I agree it is unnecessary for a single line comment.

tomgco commented:
This is a duplicate of #3823, closing!
